test/integration/garbage_collector_test.go

```diff
@@ -279,3 +283,86 @@ func TestCreateWithNonExisitentOwner(t *testing.T) {
 		t.Fatalf("deleted unexpected pod: %v", *deletedPod)
 	}
 }
+
+func createRCsPods(t *testing.T, clientSet clientset.Interface, id int, rcUIDs []types.UID, wg *sync.WaitGroup) {
+	defer wg.Done()
+	rcClient := clientSet.Core().ReplicationControllers(framework.TestNS)
+	podClient := clientSet.Core().Pods(framework.TestNS)
+	rcName := toBeDeletedRCName + strconv.Itoa(id)
+	toBeDeletedRC, err := rcClient.Create(newOwnerRC(rcName))
+	if err != nil {
+		t.Fatalf("Failed to create replication controller: %v", err)
+	}
+	rcUIDs[id] = toBeDeletedRC.ObjectMeta.UID
+	// these pods should be cascadingly deleted.
+	for j := 0; j < 3; j++ {
+		podName := garbageCollectedPodName + strconv.Itoa(id) + strconv.Itoa(j)
+		pod := newPod(podName, []v1.OwnerReference{{UID: toBeDeletedRC.ObjectMeta.UID, Name: rcName}})
+		_, err = podClient.Create(pod)
+		if err != nil {
+			t.Fatalf("Failed to create Pod: %v", err)
+		}
+	}
+}
+
+func verifyObjectsNumber(t *testing.T, clientSet clientset.Interface, rcNum, podNum int) {
+	rcClient := clientSet.Core().ReplicationControllers(framework.TestNS)
+	podClient := clientSet.Core().Pods(framework.TestNS)
+	pods, err := podClient.List(api.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to list pods: %v", err)
+	}
+	if len(pods.Items) != podNum {
+		t.Errorf("Expect %d pod(s), but got %#v", podNum, pods)
+	}
+	rcs, err := rcClient.List(api.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to list replication controllers: %v", err)
+	}
+	if len(rcs.Items) != rcNum {
+		t.Errorf("Expect %d replication controller(s), but got %#v", rcNum, rcs)
+	}
+}
+
+// This stress test the garbage collector
+func TestStressingCascadingDeletion(t *testing.T) {
+	gc, clientSet := setup(t)
+	var wg sync.WaitGroup
+	const collections = 10
+	rcUIDs := make([]types.UID, collections)
+	wg.Add(collections)
+	for i := 0; i < collections; i++ {
+		go createRCsPods(t, clientSet, i, rcUIDs, &wg)
+	}
+	wg.Wait()
+	verifyObjectsNumber(t, clientSet, collections, collections*3)
+	stopCh := make(chan struct{})
+	go gc.Run(5, stopCh)
+	defer close(stopCh)
+	// The exact duration doesn't matter. We want to delete some RCs before the
+	// garbage collector has built up the owner-dependent graph.
+	time.Sleep(10 * time.Second)
+	// delete the replication controllers
+	rcClient := clientSet.Core().ReplicationControllers(framework.TestNS)
+	for i := 0; i < collections; i++ {
+		rcName := toBeDeletedRCName + strconv.Itoa(i)
+		if err := rcClient.Delete(rcName, nil); err != nil {
+			t.Fatalf("failed to delete replication controller: %v", err)
+		}
+	}
+
+	// wait for the garbage collector to drain its queue
+	if err := wait.Poll(10*time.Second, 300*time.Second, func() (bool, error) {
+		return gc.QueuesDrained(), nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// wait for some extra time to allow the garbage collector to process all
+	// the items.
+	time.Sleep(10 * time.Second)
```

Comment:
I guess that's better than nothing.

On Thu, May 12, 2016 at 3:38 PM, Chao Xu <notifications@github.com> wrote:

> In test/integration/garbage_collector_test.go
> <https://github.com/kubernetes/kubernetes/pull/24509#discussion_r63109144>
> :
>
> > +		rcName := toBeDeletedRCName + strconv.Itoa(i)
> > +		if err := rcClient.Delete(rcName, nil); err != nil {
> > +			t.Fatalf("failed to delete replication controller: %v", err)
> > +		}
> > +	}
> > +
> > +	// wait for the garbage collector to drain its queue
> > +	if err := wait.Poll(10*time.Second, 300*time.Second, func() (bool, error) {
> > +		return gc.QueuesDrained(), nil
> > +	}); err != nil {
> > +		t.Fatal(err)
> > +	}
> > +
> > +	// wait for some extra time to allow the garbage collector to process all
> > +	// the items.
> > +	time.Sleep(10 * time.Second)
>
> How about polling over the verification function on the next line?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24509/files/39ae9bee7e17a5d30f654a1d03879a3960983dae..d897a0b82d0839ed37bb6fdc79d1a5dfa41ce0c7#r63109144>
>